Sources/XCTest/Public/XCTestCase.swift

```diff
@@ -108,41 +110,130 @@ open class XCTestCase: XCTest {
     /// - Note: Like the designated initializer for Apple XCTest's XCTestCase,
     ///   `-[XCTestCase initWithInvocation:]`, it's rare for anyone outside of
     ///   XCTest itself to call this initializer.
-    public required init(name: String, testClosure: @escaping XCTestCaseClosure) {
+    @available(*, deprecated, message: "Use init(name:closureType:) instead.")
+    public init(name: String, testClosure: @escaping XCTestCaseClosure) {
         _name = "\(type(of: self)).\(name)"
-        self.testClosure = testClosure
+        self.testClosure = .sync(testClosure)
     }
 
+    public required init(name: String, closureType: ClosureType) {
+        _name = "\(type(of: self)).\(name)"
+        self.testClosure = closureType
+    }
+    
     /// Invoking a test performs its setUp, invocation, and tearDown. In
     /// general this should not be called directly.
     open func invokeTest() {
-        performSetUpSequence()
+        if skip == nil {
+            performSetup()
+            performTest()
+            performTeardown()
+        }
```

Comment:
Two related comments here:
- I think we should more narrowly focus this `if skip == nil` guard and have it only surround the call to `performTest`, not `performSetup` or `performTeardown`.
- I think we should remove the `var shouldInvokeTestClosure = true` stored property below entirely, and instead let the nullability of `self.skip` decide whether the test closure gets invoked.

The `shouldInvokeTestClosure` Bool property is basically redundant, in other words, because in the scenarios where it is set, `self.skip` will _also_ already be set to a non-nil value. Additionally, there is some logic elsewhere which checks `if error.xct_shouldSkipTestInvocation` and decides whether or not an Error thrown during a set up method should cause the test closure to be skipped. We should preserve that behavior, but currently the `shouldInvokeTestClosure` property gets unconditionally set to `true` upon any Error being caught

Reply:
That all makes sense. I've taken care of this in the latest diff. Check it out! :)